File: server/tests/_render_html.py
"""Tiny DOM + active-markup scanner for the renderer tests (stdlib html.parser only)."""

from __future__ import annotations

import re
from collections.abc import Callable, Iterator
from dataclasses import dataclass, field
from html.parser import HTMLParser

_VOID = frozenset({"area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta", "source", "track", "wbr"})
_FORBIDDEN = frozenset({"iframe", "frame", "frameset", "object", "embed", "applet", "form", "base", "style", "template"})
_ASSET_JS = re.compile(r"assets/page\.[0-9a-f]{12}\.js")
_ASSET_CSS = re.compile(r"assets/(?:site|theme)\.[0-9a-f]{12}\.css")
_WEB = re.compile(r"(?:https?://|mailto:)", re.IGNORECASE)


@dataclass
class Node:
    tag: str
    attrs: dict[str, str | None]
    children: list[Node | str] = field(default_factory=list)

    def iter(self) -> Iterator[Node]:
        yield self
        for child in self.children:
            if isinstance(child, Node):
                yield from child.iter()

    def find_all(self, match: Callable[[Node], bool]) -> list[Node]:
        return [node for node in self.iter() if match(node)]

    def by_id(self, element_id: str) -> Node:
        found = self.find_all(lambda n: n.attrs.get("id") == element_id)
        assert len(found) == 1, f"expected one #{element_id}, found {len(found)}"
        return found[0]

    def with_class(self, name: str) -> list[Node]:
        return self.find_all(lambda n: name in (n.attrs.get("class") or "").split())

    def text(self) -> str:
        return "".join(child if isinstance(child, str) else child.text() for child in self.children)


class _Builder(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.root = Node("#document", {})
        self.stack = [self.root]

    def handle_starttag(self, tag, attrs):
        node = Node(tag, dict(attrs))
        self.stack[-1].children.append(node)
        if tag not in _VOID:
            self.stack.append(node)

    def handle_startendtag(self, tag, attrs):
        self.stack[-1].children.append(Node(tag, dict(attrs)))

    def handle_endtag(self, tag):
        for depth in range(len(self.stack) - 1, 0, -1):
            if self.stack[depth].tag == tag:
                del self.stack[depth:]
                return

    def handle_data(self, data):
        self.stack[-1].children.append(data)


def parse(html: str) -> Node:
    builder = _Builder()
    builder.feed(html)
    builder.close()
    return builder.root


def active_markup(html: str, *, pages: frozenset[str] = frozenset(), photos: frozenset[str] = frozenset()) -> list[str]:
    """Everything in `html` that could run script, load or submit elsewhere, or leave the site.

    Allowed: the page script and stylesheets under assets/, web/mail/#anchor links, links to the
    sibling files in `pages`, images of the photos in `photos`.
    """
    problems = []
    for node in parse(html).iter():
        tag, attrs = node.tag, node.attrs
        for name in attrs:
            if name.startswith("on") or name in ("style", "srcdoc", "formaction", "xlink:href"):
                problems.append(f"<{tag} {name}>")
        if tag in _FORBIDDEN:
            problems.append(f"<{tag}>")
        elif tag == "script":
            if not _ASSET_JS.fullmatch(attrs.get("src") or ""):
                problems.append(f"<script src={attrs.get('src')!r}>")
        elif tag == "link":
            if not (attrs.get("rel") == "stylesheet" and _ASSET_CSS.fullmatch(attrs.get("href") or "")):
                problems.append(f"<link {attrs!r}>")
        elif tag == "meta":
            if (attrs.get("http-equiv") or "").lower() == "refresh":
                problems.append("<meta refresh>")
        elif tag == "img":
            if (attrs.get("src") or "") not in {f"photos/{name}" for name in photos}:
                problems.append(f"<img src={attrs.get('src')!r}>")
        elif tag in ("a", "area"):
            href = attrs.get("href")
            if href is not None and not (href.startswith("#") or _WEB.match(href) or href.split("#")[0] in pages):
                problems.append(f"<a href={href!r}>")
        elif "src" in attrs or "href" in attrs or "action" in attrs or "data" in attrs:
            problems.append(f"<{tag} {attrs!r}>")
    return problems

File: desktop/tests/core/renderHtml.js
// Tiny DOM + active-markup scanner for the renderer tests.

const VOID = new Set(['area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'link', 'meta', 'source', 'track', 'wbr']);
const FORBIDDEN = new Set(['iframe', 'frame', 'frameset', 'object', 'embed', 'applet', 'form', 'base', 'style', 'template']);
const ASSET_JS = /^assets\/page\.[0-9a-f]{12}\.js$/;
const ASSET_CSS = /^assets\/(?:site|theme)\.[0-9a-f]{12}\.css$/;
const WEB = /^(?:https?:\/\/|mailto:)/i;

class Node {
  constructor(tag, attrs) {
    this.tag = tag;
    this.attrs = attrs;
    this.children = [];
  }

  *iter() {
    yield this;
    for (const child of this.children) {
      if (child instanceof Node) yield* child.iter();
    }
  }

  findAll(match) {
    return [...this.iter()].filter(match);
  }

  byId(elementId) {
    const found = this.findAll((n) => n.attrs.id === elementId);
    if (found.length !== 1) throw new Error(`expected one #${elementId}, found ${found.length}`);
    return found[0];
  }

  text() {
    return this.children.map((child) => (typeof child === 'string' ? child : child.text())).join('');
  }
}

function decodeEntities(value) {
  return String(value).replace(/&(#x[0-9a-f]+|#\d+|[a-z]+);/gi, (entity, body) => {
    const lower = body.toLowerCase();
    if (lower === 'lt') return '<';
    if (lower === 'gt') return '>';
    if (lower === 'amp') return '&';
    if (lower === 'quot') return '"';
    if (lower === 'apos') return "'";
    if (lower === 'minus') return '\u2212';
    if (body[0] === '#') {
      const code = body[1] === 'x' || body[1] === 'X' ? parseInt(body.slice(2), 16) : Number(body.slice(1));
      if (Number.isFinite(code)) return String.fromCharCode(code);
    }
    return entity;
  });
}

function parseAttrs(raw) {
  const attrs = {};
  const re = /([^\s=]+)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s"'>=]+)))?/g;
  let match;
  while ((match = re.exec(raw))) {
    const value = match[2] !== undefined ? match[2] : match[3] !== undefined ? match[3] : match[4] !== undefined ? match[4] : null;
    attrs[match[1]] = value === null ? null : decodeEntities(value);
  }
  return attrs;
}

function parse(html) {
  const root = new Node('#document', {});
  const stack = [root];
  const re = /<!--[\s\S]*?-->|<!DOCTYPE[^>]*>|<\/([a-zA-Z][\w:-]*)\s*>|<([a-zA-Z][\w:-]*)([^>]*)>/g;
  let last = 0;
  let match;
  while ((match = re.exec(html))) {
    if (match.index > last) stack[stack.length - 1].children.push(decodeEntities(html.slice(last, match.index)));
    last = re.lastIndex;
    if (match[0].startsWith('<!--') || match[0].startsWith('<!')) continue;
    if (match[1]) {
      const tag = match[1].toLowerCase();
      for (let depth = stack.length - 1; depth > 0; depth -= 1) {
        if (stack[depth].tag === tag) {
          stack.length = depth;
          break;
        }
      }
      continue;
    }
    const tag = match[2].toLowerCase();
    let attrSource = match[3] || '';
    const selfClose = attrSource.trim().endsWith('/');
    if (selfClose) attrSource = attrSource.replace(/\/\s*$/, '');
    const node = new Node(tag, parseAttrs(attrSource));
    stack[stack.length - 1].children.push(node);
    if (!VOID.has(tag) && !selfClose) stack.push(node);
  }
  if (last < html.length) stack[stack.length - 1].children.push(decodeEntities(html.slice(last)));
  return root;
}

function activeMarkup(html, { pages = new Set(), photos = new Set() } = {}) {
  const pageSet = pages instanceof Set ? pages : new Set(pages);
  const photoSet = photos instanceof Set ? photos : new Set(photos);
  const problems = [];
  for (const node of parse(html).iter()) {
    const { tag, attrs } = node;
    for (const name of Object.keys(attrs)) {
      if (name.startsWith('on') || name === 'style' || name === 'srcdoc' || name === 'formaction' || name === 'xlink:href') {
        problems.push(`<${tag} ${name}>`);
      }
    }
    if (FORBIDDEN.has(tag)) {
      problems.push(`<${tag}>`);
    } else if (tag === 'script') {
      if (!ASSET_JS.test(attrs.src || '')) problems.push(`<script src=${JSON.stringify(attrs.src)}>`);
    } else if (tag === 'link') {
      if (!(attrs.rel === 'stylesheet' && ASSET_CSS.test(attrs.href || ''))) problems.push(`<link ${JSON.stringify(attrs)}>`);
    } else if (tag === 'meta') {
      if ((attrs['http-equiv'] || '').toLowerCase() === 'refresh') problems.push('<meta refresh>');
    } else if (tag === 'img') {
      if (![...photoSet].map((name) => `photos/${name}`).includes(attrs.src || '')) {
        problems.push(`<img src=${JSON.stringify(attrs.src)}>`);
      }
    } else if (tag === 'a' || tag === 'area') {
      const href = attrs.href;
      if (href !== undefined && !(href.startsWith('#') || WEB.test(href) || pageSet.has(href.split('#')[0]))) {
        problems.push(`<a href=${JSON.stringify(href)}>`);
      }
    } else if ('src' in attrs || 'href' in attrs || 'action' in attrs || 'data' in attrs) {
      problems.push(`<${tag} ${JSON.stringify(attrs)}>`);
    }
  }
  return problems;
}

module.exports = { Node, parse, activeMarkup };
